applications/FemToDemApplication/custom_processes/compute_normalized_free_energy_on_nodes_process.cpp

```diff
@@ -0,0 +1,335 @@
+    if (mComputeNormalizedFreeEnergy) {
+        const double fracture_energy_tension = rMatProps[FRAC_ENERGY_T];
+        const double yield_tension = rMatProps[YIELD_STRESS_T];
+        const double yield_compression = rMatProps[YIELD_STRESS_C];
+        const double ratio = yield_compression / yield_tension;
+        const double fracture_energy_compression = fracture_energy_tension * std::pow(ratio, 2.0);
+        const double density = rMatProps[DENSITY];
+        
+        KRATOS_ERROR_IF(yield_tension < tolerance) << "The YIELD_TENSION is null or not assigned" << std::endl;
+        KRATOS_ERROR_IF(yield_compression < tolerance) << "The YIELD_COMPRESSION is null or not assigned" << std::endl;
+        KRATOS_ERROR_IF(fracture_energy_tension < tolerance) << "The FRACTURE ENERGY is null or not assigned" << std::endl;
+        KRATOS_ERROR_IF(density < tolerance) << "The DENSITY is null or not assigned" << std::endl;
+
+        double characteristic_length, r;
+        if (mDimension == 2) {
+            characteristic_length = this->CalculateCharacteristicLength2D(rGeometry);
+            r = this->ComputeTensionFactor2D(rStressVector);
+        } else {
+            characteristic_length = this->CalculateCharacteristicLength3D(rGeometry);
+            r = this->ComputeTensionFactor3D(rStressVector);
+        }
+        const double g_t = fracture_energy_tension / characteristic_length;
+        const double g_c = fracture_energy_compression / characteristic_length;
+        normalized_free_energy *= (1.0 - Damage);
+        normalized_free_energy /= (2.0 * density);
+        normalized_free_energy *= (r / g_t + (1.0 - r) / g_c);
+    }
+    return normalized_free_energy;
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+double ComputeNormalizedFreeEnergyOnNodesProcess::CalculateCharacteristicLength2D(
+    const Geometry<Node<3>>& rGeometry
+    )
+{
+    Vector node_1_coordinates = rGeometry[0].Coordinates();
+    Vector node_2_coordinates = rGeometry[1].Coordinates();
+    Vector node_3_coordinates = rGeometry[2].Coordinates();
+
+    const double length1 = norm_2(node_1_coordinates-node_2_coordinates);
+    const double length2 = norm_2(node_2_coordinates-node_3_coordinates);
+    const double length3 = norm_2(node_3_coordinates-node_1_coordinates);
+    return (length1 + length2 + length3) / 3.0;
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+double ComputeNormalizedFreeEnergyOnNodesProcess::CalculateCharacteristicLength3D(
+    Geometry<Node<3>>& rGeometry
+    )
+{
+    Vector lengths = ZeroVector(6);
+    auto r_edges = rGeometry.Edges();
+    for (unsigned int edge = 0; edge < 6; edge++) { // Loop over edges
+        const double X1 = r_edges[edge][0].X0();
```

Comment:
There is a method called InitialCoordinates or InitialPosition, will avoid to access several times to the database